### ISSUE TITLE
Snippets and suggestions to slow

### ISSUE THREAD

**Issue by trahibidadido** [tsufeki/tenkawa-php-language-server#21]
After installing vscode ext basic snippets (fun, class, con etc.) takes 3-5sec to apply.

Any suggestions?

**Comment by tsufeki** [comment]
Do you mean the time to show the autocompletion dropdown? vscode seems to wait for responses from all servers before displaying anything, and tenkawa's semantic completion is pretty slow (especially in big projects with a lot of dependencies). I'm trying to come up with ideas to improve it, but this is somewhat difficult problem, sorry. One thing to make sure is if indexing has finished (the popup "Indexing files..." disappeared).

**Comment by xiaohuilam** [comment]
This issue, i think, is about the project's protocol structures.

PHP is a synchronized language, it means when a `request` (sort of, maybe not correctly at all) is under processing, in this thread, it can not accept next request same time.

To solve this problem, seems need package owner to implement coroutine functions, or just fork a new thread when a request received (cpu might overweight?).

**Comment by MarTango** [comment]
React\EventLoop is a dependency, which suggests that tenkawa does run asynchronously. (I'm not entirely sure though, as I haven't looked at the source code in depth).

**Comment by tsufeki** [comment]
Yes the server runs on coroutines (react & recoil) but most of the task are cpu-bound and often hard to split because they call synchronous code (like phpstan). Also this is the reason that implementing cancellations would not be very useful right now.

Starting several processes would be complicated as they'd need to share a lot of state (open documents, index, caches etc.)

**Comment by xiaohuilam** [comment]
@tsufeki
strongly advice to explore with https://github.com/swoole/swoole-src, it says provides a sync to async quick magic calls `Swoole\Runtime::enableCoroutine()`, i'm sure asynchronization works fine with pdo/curl because i have private project is powered by it, but i am not sure if it work with file r/w io...
